loader/loader.c

```diff
@@ -4129,49 +4468,144 @@ void loader_deactivate_layers(const struct loader_instance *instance, struct loa
+            if (vk_inst_layers == NULL) {
+                res = VK_ERROR_OUT_OF_HOST_MEMORY;
+                goto out;
+            }
+            for (uint32_t cur_layer = 0; cur_layer < separator_count; ++cur_layer) {
+                vk_inst_layers[cur_layer] =
+                    loader_instance_heap_calloc(inst, VK_MAX_EXTENSION_NAME_SIZE, VK_SYSTEM_ALLOCATION_SCOPE_COMMAND);
+                if (vk_inst_layers[cur_layer] == NULL) {
+                    res = VK_ERROR_OUT_OF_HOST_MEMORY;
+                    goto out;
+                }
+            }
+
+            strcpy(name, layer_env);
+
+            loader_log(inst, VULKAN_LOADER_WARN_BIT | VULKAN_LOADER_LAYER_BIT, 0, "env var \'%s\' defined and adding layers \"%s\"",
+                       env_name, name);
+
+            // First look for the old-fashion layers forced on with VK_INSTANCE_LAYERS
+            while (name && *name) {
+                next = loader_get_next_path(name);
+
+                if (strlen(name) > 0) {
+                    strncpy(vk_inst_layers[vk_inst_layer_count++], name, VK_MAX_EXTENSION_NAME_SIZE);
+                }
+                name = next;
+            }
+        }
     }
-    strcpy(name, layer_env);
 
-    loader_log(inst, VULKAN_LOADER_WARN_BIT | VULKAN_LOADER_LAYER_BIT, 0,
-               "loader_add_environment_layers: Env Var %s defined and adding layers %s", env_name, name);
+    // Loop through all the layers and check the enable/disable filters as well as the VK_INSTANCE_LAYERS value.
+    for (uint32_t i = 0; i < source_list->count; i++) {
+        struct loader_layer_properties *source_prop = &source_list->list[i];
 
-    while (name && *name) {
-        next = loader_get_next_path(name);
-        res = loader_add_layer_name_to_list(inst, name, type_flags, source_list, target_list, expanded_target_list);
-        if (res != VK_SUCCESS) {
-            goto out;
+        // If it doesn't match the type, or the name isn't what we're looking for, just continue
+        if ((source_prop->type_flags & type_flags) != type_flags) {
+            continue;
+        }
+
+        // We found a layer we're interested in, but has it been disabled...
+        bool adding = true;
+        bool is_implicit = (0 == (source_prop->type_flags & VK_LAYER_TYPE_FLAG_EXPLICIT_LAYER));
+        bool disabled_by_type = (is_implicit) ? (NULL != disable_filter && disable_filter->disable_all_implicit)
+                                              : (NULL != disable_filter && disable_filter->disable_all_explicit);
+        if ((NULL != disable_filter &&
+             (disable_filter->disable_all || disabled_by_type ||
+              check_name_matches_filter(inst, source_prop->info.layerName, &disable_filter->additional_filters))) &&
+            (NULL == enable_filter || !check_name_matches_filter(inst, source_prop->info.layerName, enable_filter))) {
+            loader_log(inst, VULKAN_LOADER_WARN_BIT | VULKAN_LOADER_LAYER_BIT, 0,
```

Comment:
Do we need to check if the enable_filter here would enable the layer if we are going to do that later down below?

Reply:
Good point.  Verified removing this code has no effect on the tests.